### ISSUE TITLE
webconnectivity: handle TLS misconfigured websites

### ISSUE THREAD

**Issue by bassosimone** [ooni/probe#2300]
Let's start from the [obligatory MAT search](https://explorer.ooni.org/chart/mat?test_name=web_connectivity&domain=www.bing-amp.com&since=2021-09-12&until=2022-09-12&axis_x=measurement_start_day):

<img width="1260" alt="Screenshot 2022-09-13 at 09 11 53" src="https://user-images.githubusercontent.com/337298/189834540-5b161b44-d99a-4ce5-966e-3b633223ac43.png">

We see that since recently we have had a growth in the fraction of anomalies and failures. To start getting some confidence about what is going on there, let's inspect three recent measurements:

1. [a successful measurement using v0.4.0](https://explorer.ooni.org/measurement/20220911T235327Z_webconnectivity_US_20115_n1_SoQ0sRqi8KhN3yTZ?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js)

2. [a failed measurement using v0.4.1](https://explorer.ooni.org/measurement/20220911T235634Z_webconnectivity_GB_2856_n1_qMwWBf6JINXcBB5g?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js)

3. [an anomalous measurement using v0.4.0](https://explorer.ooni.org/measurement/20220911T235219Z_webconnectivity_GB_5378_n1_P924vqMydUT20793?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js)

### First measurement

The first measurement of the pack clearly failed inside the control:

<img width="502" alt="Screenshot 2022-09-13 at 09 19 19" src="https://user-images.githubusercontent.com/337298/189836172-7cf91018-a4bc-4a50-84d2-6beec45d027a.png">

However, the algorithm for computing the summary determines that all is good:

<img width="351" alt="Screenshot 2022-09-13 at 09 20 07" src="https://user-images.githubusercontent.com/337298/189836324-dccda08f-28b3-4fa8-8dde-830b41e45541.png">

Why is that? Is this a bug in the probe? Well, if we look at the TLS handshakes for the probe we see:

<img width="605" alt="Screenshot 2022-09-13 at 09 29 47" src="https://user-images.githubusercontent.com/337298/189838325-6e4e2c82-3708-4527-bdce-817efb83be31.png">

So, it seems _in this case_ we used an IP address where we can TLS handshake. Also, if we look at HTTPS:

<img width="377" alt="Screenshot 2022-09-13 at 09 30 52" src="https://user-images.githubusercontent.com/337298/189838580-7a4f2457-3ac6-4fe5-8082-4e96667696ec.png">

Which means the HTTPS request did not fail.

So, now it should be more clear what we see if we notice that `x_status` is `1`, which means `StatusSuccessSecure` (see [i/e/e/w/summary.go](https://github.com/ooni/probe-cli/blob/cf6dbe48e028826ab493794ef7d7d32a4bee3f14/internal/engine/experiment/webconnectivity/summary.go)). This state is an optimization inside v0.4 where we always consider successful a measurement using HTTPS where we can perform a TLS handshake and fetch data _regardless_ of what happened in the TH. 

### Second measurement

The second measurement has `x_status` equal to `16`, which means `StatusAnalysisControlFailure` (see [i/e/e/w/summary.go](https://github.com/ooni/probe-cli/blob/cf6dbe48e028826ab493794ef7d7d32a4bee3f14/internal/engine/experiment/webconnectivity/summary.go)). So, this seems also an acceptable outcome. Because the control is failed, we flag the measurement as errored. We could of course do better, like recognizing that the website is down if both the control and the probe failed in the same way.

### Third measurement

This measurement actually looks good. The control measurement succeeded (WTF?!) while the probe failed because of `ssl_invalid_hostname`. So, the measurement itself is flagged as anomaly. And, yeah, this feels like an anomaly.

### Preliminary conclusion for v0.4

I think this issue is worth more investigation, but time is limited. So, for now, let's spell out a preliminary conclusion.

We think that what is happening here is that globally only _some_ addresses are configured to correctly use the SNI that lives inside the URL. This means that probes and THs could fail or succeed depending on what happens to them, which in turn depends on which connections the DNS gives to them. I think all the probes here behave reasonably, given the circumstances.

It would be cool to introduce inside v0.5 a state where we recognize the TLS is misconfigured for both the probe and the TH and hence say something along the lines of "website down".

### v0.5 measurements

We have fewer v0.5 measurements at the moment. [Here's one](https://explorer.ooni.org/measurement/20220911T105037Z_webconnectivity_IT_30722_n1_ruzuQ219SmIO9SrT?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js) run by me a few days ago.

This measurement falls into the case where both the TH and the probe fail. So, it's a good candidate for detecting this corner case and declaring that the website is down because of TLS misconfiguration.

#### First measurement

The [first v0.5 measurement](https://explorer.ooni.org/measurement/20220913T080946Z_webconnectivity_IT_30722_n1_vvTeJN1bKHYsoLBY?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js) I'd like to show you is a case of anomaly. For some reason (maybe my IP address changed at last?) I am now using `1.th.ooni.org` rather than `0.th.ooni.org` by default.

So, in the control the TLS handshake succeeds:

<img width="441" alt="Screenshot 2022-09-13 at 10 15 28" src="https://user-images.githubusercontent.com/337298/189848516-1240fbb8-a6f1-4e21-b2ad-07984742f862.png">

However, the probe sees this:

<img width="423" alt="Screenshot 2022-09-13 at 10 16 33" src="https://user-images.githubusercontent.com/337298/189848757-3e364129-e589-45d5-bb8b-adaa5e450360.png">

And so it seems justified to say:

```
[      2.712213] <warn> TLS: endpoint 204.79.197.219:443 is blocked (see #4): ssl_invalid_hostname
[      2.712223] <warn> ANOMALY: flags=4 accessible=false, blocking=http-failure
```

#### Second measurement

I ran [another measurement](https://explorer.ooni.org/measurement/20220913T081222Z_webconnectivity_IT_30722_n1_HlzOA3E7VlBCbFlB?input=https%3A%2F%2Fwww.bing-amp.com%2Fv0%2Famp-iframe-0.1.js) forcing the backend to be `0.th`. 

Here's the control part of the TLS handshake:

<img width="451" alt="Screenshot 2022-09-13 at 10 18 37" src="https://user-images.githubusercontent.com/337298/189849183-c545e6c1-279b-4897-907d-fe722f622db4.png">

Here's the probe's part:

<img width="391" alt="Screenshot 2022-09-13 at 10 19 08" src="https://user-images.githubusercontent.com/337298/189849324-738817bd-98d6-4509-8349-ff8751997f9a.png">

So, with the new patch I'm working on (and will commit soon) we obtain:

```
[      0.977307] <info> website likely down: all TLS handshake attempts failed for both probe and TH
[      0.977323] <info> WEBSITE_DOWN_TLS: flags=0, accessible=false, blocking=false
```

### Conclusion

We can surely mitigate the issue on the probe side by saying "website down" rather than "measurement failed" in the cases in which it makes sense to do so. I am still a bit uneasy about the same TCP endpoint providing different results regardless of what TH is measuring, which probably hints at anycast or region specific configurations. What I am trying to say here is that this `bing-amp.com` URL inside the test list will increase the noise we see. More noise creates more possibility for errors when analyzing the data. It would be interesting to discuss with @hellais whether and how this could change if we adopt the https://github.com/ooni/data approach as a foundation for measurements.

**Comment by bassosimone** [comment]
We already have an issue where we describe hopping between failure and success depending on the TH we use: https://github.com/ooni/probe/issues/2298.

We also have fixed the issue for webconnectivity v0.5.

So, I am going to flag this issue as fixed for webconnectivity v0.5 and we'll close this issue when all users will be using this version of webconnectivity.

**Comment by bassosimone** [comment]
![pRr9-V](https://github.com/ooni/probe/assets/337298/283df799-f7de-4903-a442-ff01cd705de8)

Okay, we have fixed this in Web Connectivity LTE, hence we can close this issue. We have a specific test case for that that runs for every commit and ensures we keep detecting this common case.
